core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -90,30 +93,239 @@
+        return;
+      }
+      computeState();
+    }
+
+    /**
+     * @return current file position
+     */
+    public long getPos() {
+      return mPos;
+    }
+
+    /**
+     * @return current remaining bytes
+     */
+    public long getRemaining() {
+      return mRemaining;
+    }
+
+    /**
+     * @return ID of the block where the current file position belongs to
+     */
+    public long getBlockId() {
+      return mBlockId;
+    }
+
+    /**
+     * @return the start position of the block
+     */
+    public long getBlockStartPos() {
+      return mBlockStartPos;
+    }
+
+    /**
+     * @return the start position of the next block
+     */
+    public long getNextBlockStartPos() {
+      return mNextBlockStartPos;
+    }
+
+    /**
+     * @return whether EOF is reached
+     */
+    public boolean isEOF() {
+      return mPos == mFileLength;
+    }
+
+    /**
+     * Sets the states to special values when EOF is reached.
+     */
+    private void setEOFState() {
+      mPos = mBlockStartPos = mNextBlockStartPos = mFileLength;
+      mRemaining = 0;
+      mBlockIndex = mBlockIds.size();
+      mBlockId = EOF_BLOCK_ID;
+    }
+
+    /**
+     * Recomputes the states based on mPos.
+     * Comparing to {@link #incrementState()}, this method has higher CPU cost, it should only be
+     * used when {@link #incrementState()} cannot be used.
+     */
+    private void computeState() {
+      if (mPos >= mFileLength) {
+        setEOFState();
+        return;
+      }
+
+      mBlockIndex = (int) (mPos / mBlockSize);
+      mBlockId = mBlockIds.get(mBlockIndex);
+
+      mBlockStartPos = mBlockIndex * mBlockSize;
+      mNextBlockStartPos = Math.min(mBlockStartPos + mBlockSize, mFileLength);
+    }
+
+    /**
+     * Assumes the position just enters the next block, updates the states.
+     * Comparing to {@link #computeState()}, this method has lower CPU cost.
+     */
+    private void incrementState() {
+      if (mPos >= mFileLength) {
+        setEOFState();
+        return;
+      }
+
+      mBlockIndex++;
+      mBlockId = mBlockIds.get(mBlockIndex);
+
+      mBlockStartPos = mNextBlockStartPos;
+      mNextBlockStartPos = Math.min(mNextBlockStartPos + mBlockSize, mFileLength);
+    }
+  }
+
+  /**
+   * @param pos the file position, assumed to be in range [0, file length)
+   * @return the block ID for pos
+   */
+  private long getBlockId(long pos) {
+    return mStatus.getBlockIds().get((int) (pos / mBlockSize));
+  }
+
+  /**
    * Creates a new file input stream.
    *
    * @param status the file status
    * @param options the client options
    * @param context file system context
    * @return the created {@link FileInStream} instance
+   * @throws IOException when failed to create a new {@link UnknownLengthFileInStream}
    */
   public static FileInStream create(URIStatus status, InStreamOptions options,
-      FileSystemContext context) {
+      FileSystemContext context) throws IOException {
```

Comment:
this is a public API change. is this required?

Reply:
normally, users should use `FileSystem.openFile` which calls `FileInStream.create`, `openFile` already throws `IOException`.

Reply:
ack